smithy-typescript-codegen/src/main/java/software/amazon/smithy/typescript/codegen/integration/HttpBindingProtocolGenerator.java

```diff
@@ -584,11 +604,105 @@ protected void serializeInputPayload(
+            }
+        });
+    }
+
+    private void generateEventStreamSerializer(GenerationContext context, MemberShape member, Shape target) {
+        TypeScriptWriter writer = context.getWriter();
+        Symbol targetSymbol = context.getSymbolProvider().toSymbol(target);
+        String serializingFunction;
+        if (target instanceof StructureShape) {
+            // Single-event event stream. Supply event itself instead of event key-value pair to event deserializer
+            serializingFunction = ProtocolGenerator.getSerFunctionName(targetSymbol, context.getProtocolName())
+                    + "_event(event[Object.keys(event)[0]], context)";
+            this.serializeEventShapes.add(target.asStructureShape().get());
+        } else if (target instanceof UnionShape) {
+            // Multi-event event stream. Save the dispatcher
+            serializingFunction = ProtocolGenerator.getSerFunctionName(targetSymbol, context.getProtocolName())
+                    + "_event(event, context)";
+            this.serializeEventUnions.add(target.asUnionShape().get());
+        } else {
+            throw new CodegenException(String.format("Unexpected shape type with eventstream trait: `%s`",
+                    target.getType()));
+        }
+        writer.openBlock("context.eventStreamMarshaller.serialize(", ");", () -> {
+            writer.write("input.$L,", member.getMemberName());
+            writer.write("event => $L", serializingFunction);
         });
     }
 
+    private void generateEventSerializer(
+            GenerationContext context,
+            StructureShape event
+    ) {
+        TypeScriptWriter writer = context.getWriter();
+        SymbolProvider symbolProvider = context.getSymbolProvider();
+        Symbol symbol = symbolProvider.toSymbol(event);
+        String methodName = ProtocolGenerator.getSerFunctionName(symbol, context.getProtocolName()) + "_event";
+        Model model = context.getModel();
+        // Handle the general response.
+        writer.openBlock("const $L = (\n"
+                        + "  input: $L,\n"
+                        + "  context: __SerdeContext\n"
+                        + "): __Message => {", "}",
+                methodName,
+                symbol.getName(),
+                () -> {
+                    writer.openBlock("const message: __Message = {", "}", () -> {
+                        writer.openBlock("headers: {", "},", () -> {
+                            //fix headers required by event stream
+                            writer.write("\":event-type\": { type: \"string\", value: \"$L\" },", symbol.getName());
+                            writer.write("\":message-type\": { type: \"string\", value: \"event\" }");
+                        });
+                        writer.write("body: new Uint8Array()");
+                    });
+                    List<MemberShape> headerMembers = event.getAllMembers().values().stream()
```

Comment:
Can these chunks of code for handling message headers and bodies be broken down in to other functions like for standard http requests?